cuqi/randomvariable/_randomvariable.py

```diff
@@ -0,0 +1,191 @@
+from __future__ import annotations
+from typing import List, Any
+from ._ast import RandomVariableNode
+from ._orderedset import OrderedSet
+import operator
+import cuqi
+
+
+class RandomVariable:
+    """ Random variable defined by a base distribution with potential algebraic operations applied to it.
+
+    The random variable class is still experimental.
+
+    The random variable can be viewed as a lazily evaluated array. This allows for the definition of
+    Bayesian models in a natural way.
+
+    Example
+    -------
+
+    Basic usage:
+
+    .. code-block:: python
+
+        from cuqi.distribution import Gaussian
+
+        x = Gaussian(0, 1)
+
+        print(x+1)  # returns a RandomVariable object with operation recorded
+
+    Directly creating random variable:
+   
+    .. code-block:: python
+
+        from cuqi.distribution import Gaussian
+
+        x = Gaussian(0, 1)
+
+        x_rv = x.as_random_variable()  # returns a RandomVariable object
+
+        print((x_rv+1)**2)  # returns a RandomVariable object with operation recorded
+
+    Defining Bayesian problem using algebra on distributions (creating random variables implicitly):
+
+    .. code-block:: python
+
+        from cuqi.testproblem import Deconvolution1D
+        from cuqi.distribution import Gaussian, Gamma, GMRF
+        from cuqi.problem import BayesianProblem
+        import numpy as np
+        A, y_obs, info = Deconvolution1D.get_components()
+
+        # Bayesian problem
+        d = Gamma(1, 1e-4)
+        s = Gamma(1, 1e-4)
+        x = GMRF(np.zeros(A.domain_dim), d)
+        y = Gaussian(A @ x, 1/s)
+
+        BP = BayesianProblem(y, x, s, d)
+        BP.set_data(y=y_obs)
+        BP.UQ()
+
+    """
+
+    def __init__(self, distributions: set, tree: RandomVariableNode = None):
+        """ Create random variable from distribution """
+        # Convert to list if single distribution
+        if not isinstance(distributions, OrderedSet):
+            distributions = OrderedSet([distributions])
+        
+        # Check that distributions have names
+        for distribution in distributions:
+            if distribution.name is None or distribution.name == "distribution":
+                raise ValueError(
+                    "Unable to create random variable from distribution without name. Ensure "
+                    "distribution is defined as a variable, e.g. x = Gaussian(0, 1) or provide "
+                    "a name, e.g. Gaussian(0, 1, name='x')"
+                )
+        
+        self._distributions = distributions
+        """ The distribution from which the random variable originates. """
```

Comment:
Just a question about using docstring here and below, is this preferred over using regular comment?

Reply:
This adds a docstring to the property similar to adding it when using @property.